sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/joins.scala

```diff
@@ -304,8 +304,8 @@ trait JoinSelectionHelper extends Logging {
       canBroadcastBySize(join.right, conf) && !hintToNotBroadcastRight(join.hint)
     }
     getBuildSide(
-      canBuildBroadcastLeft(join.joinType) && buildLeft,
```

Comment:
This PR reminds me that `buildLeft` is not actually cheap to calculate, likely more expensive than `canBuildBroadcastLeft` invocation.

Shall we turn `buildLeft/buildRight` into lazy val and revert the order here?

Reply:
Okay, let me try it.